README.md
# link_preview

Dart library to get information needed to load previews for links

## Getting Started

### Adding dependency
Add "link_preview: {{latest-version}}" under dependencies in your pubspec.yaml.
Perform "Pub get" after adding dependency.

### Warm up link preview

Apps can warm up link_preview on app launch. Warm up API will cache data needed to extract page info.
Warm up is an optional step. Warming up will help reduce 1-2 seconds on first page load.

```
    LinkPreview.warmUp();
```

### Loading Page Info to show preview

```
    var data = await LinkPreview.getPageInfo(url);
    print(data["provider_name"]);
    print(data["thumbnail_url"]);
    print(data["provider_url"]);
```
data will be in same specification as OEmbed provider.
Ref. https://oembed.com/#section2


CHANGELOG.md
## [0.0.1] - 21/7/2020

* Load Page info from OEmbed provider


test/linkpreview_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:link_preview/link_preview.dart';

var links = [
  "https://www.circuitlab.com/circuit/pz43c8wwj2de/potentiometer-inquiry/",
  "https://coub.com/view/2gkzwz",
  "https://www.facebook.com/100034183263437/videos/302917017524405/",
  "https://www.iheart.com/podcast/1119-red-table-talk-63922369/episode/motherhood-68561795/?embed=true",
  "https://youtu.be/xaFPt4qwZOI"
];

void loadData(String url) async {
  DateTime startTime = DateTime.now();
  var data = await LinkPreview.getPreviewInfo(url);
//  print(data);
  print(
      "${DateTime.now().millisecondsSinceEpoch - startTime.millisecondsSinceEpoch} ms to load $url");
}

void main() {
  test('Test: Warm up API', () async {
    DateTime startTime = DateTime.now();
    LinkPreview.warmUp();
    print(
        "Warm up initiated in ${DateTime.now().millisecondsSinceEpoch - startTime.millisecondsSinceEpoch}");

    await Future.delayed(Duration(milliseconds: 1000), () {
      // Waiting for scheme to load
    });

    startTime = DateTime.now();
    var data = await LinkPreview.getPreviewInfo(links[0]);
    print(
        "${DateTime.now().millisecondsSinceEpoch - startTime.millisecondsSinceEpoch} ms to load ${links[0]}");

    startTime = DateTime.now();
    data = await LinkPreview.getPreviewInfo(links[1]);
    print(
        "${DateTime.now().millisecondsSinceEpoch - startTime.millisecondsSinceEpoch} ms to load ${links[1]}");

    startTime = DateTime.now();
    data = await LinkPreview.getPreviewInfo(links[2]);
    print(
        "${DateTime.now().millisecondsSinceEpoch - startTime.millisecondsSinceEpoch} ms to load ${links[2]}");

    startTime = DateTime.now();
    data = await LinkPreview.getPreviewInfo(links[3]);
    print(
        "${DateTime.now().millisecondsSinceEpoch - startTime.millisecondsSinceEpoch} ms to load ${links[3]}");

    startTime = DateTime.now();
    data = await LinkPreview.getPreviewInfo(links[4]);
    print(
        "${DateTime.now().millisecondsSinceEpoch - startTime.millisecondsSinceEpoch} ms to load ${links[4]}");
  });

  test('Test: Warm up API', () async {
    print("Test: Serial API calls without warm up");
    loadData(links[0]);
    loadData(links[1]);
    loadData(links[2]);
    loadData(links[3]);
    loadData(links[4]);

    await Future.delayed(Duration(milliseconds: 5000), () {
      // Waiting for scheme to load
    });
  });
}


lib/link_preview.dart
library link_preview;
import 'package:http/http.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';

class LinkPreview {

  static final LinkPreview _singleton = LinkPreview._internal();

  factory LinkPreview() => _singleton;

  LinkPreview._internal();

  static Map<String, String> _schemeMap = Map();

  static Future<Map> _cacheOEmbedSchemeMap() async {
    if (_schemeMap.length > 0) {
      return _schemeMap;
    }
    try {
      _schemeMap.clear();
      Response response = await http.get("https://oembed.com/providers.json");
      if (_schemeMap.length > 0) {
        return _schemeMap;
      }

      List<dynamic> items = jsonDecode(response.body);

      for (dynamic item in items) {
        try {
          var endpoints = item["endpoints"];
          String providerUrl = item["provider_url"];
          for (dynamic endpoint in endpoints) {
            var schemes = endpoint["schemes"];
            String url = endpoint["url"];
            url = url.replaceAll("{format}", "json");
            if (schemes == null) {
              if (!url.endsWith(".xml")) {
                _schemeMap[providerUrl + ".+"] = url;
              }
              continue;
            }
            for (String scheme in schemes) {
              scheme = scheme.replaceAll("*", ".+");
              scheme = scheme.replaceAll("http://", "https://");
              _schemeMap[scheme] = url;
            }
          }
        } catch (e) {
          print(e);
        }
      }
    } catch (e) {
      print(e);
    }
    return _schemeMap;
  }

  /// Apps can warm up link_preview on app launch using this API.
  /// Warm up API will cache data needed to extract page info.
  /// Warm up is an optional step. Warming up will help reduce 1-2 seconds on first page load.
  static Future<void> warmUp() async {
    await _cacheOEmbedSchemeMap();
  }

  /// This API return page information if the provided url.
  /// Page information will be in OEmbed response structure.
  /// Ref. https://oembed.com/#section2
  ///
  /// Returns null if there is error in loading page information
  static Future<dynamic> getPageInfo(String url) async {
    try {
      Map<String, String> schemeMap = await _cacheOEmbedSchemeMap();

      String oembedUrl;
      schemeMap.forEach((key, value) {
        RegExp exp = new RegExp(key);

        Iterable<Match> matches = exp.allMatches(url);
        if (matches.length > 0) {
          oembedUrl = "$value?format=json&url=$url";
        }
      });

      if (oembedUrl == null) {
        print("Match not found for $url");
        return null;
      }

      Response response = await http.get(oembedUrl);
      return jsonDecode(response.body);
    } catch (e) {
      print(e);
    }
  }
}


example/main.dart
import 'package:link_preview/link_preview.dart';

var links = [
  "https://www.circuitlab.com/circuit/pz43c8wwj2de/potentiometer-inquiry/",
  "https://coub.com/view/2gkzwz",
  "https://www.iheart.com/podcast/1119-red-table-talk-63922369/episode/motherhood-68561795/?embed=true",
  "https://youtu.be/xaFPt4qwZOI"
];

void main(List<String> arguments) async {
  // Normal sample
  fetchPageInfoDirectly();

  // Warm up example
  warmUpAndFetchPageInfo();
}

void warmUpAndFetchPageInfo() async {
  // Apps can warm up link_preview on app launch
  // Warm up API will cache data needed to extract page info
  // Warm up is an optional step. Warming up will help reduce 1-2 seconds
  LinkPreview.warmUp();

  await Future.delayed(Duration(milliseconds: 1000), () {
    // Waiting for scheme to load
  });

  for (String url in links) {
    var data = await LinkPreview.getPageInfo(url);
    print(data["thumbnail_url"]);
    print(data["provider_url"]);
  }
}

void fetchPageInfoDirectly() async {
  for (String url in links) {
    var data = await LinkPreview.getPageInfo(url);
    print(data["provider_name"]);
  }
}
